### ISSUE TITLE
Name of first argument in metaclass method

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#451]
**Describe the bug**
Pyright reports that "instance methods should take 'self' as first argument" but I'm not sure if there is a strong enough tradition in Python to name the first argument of a _metaclass_ instance method 'self' - in no small part because of how obscure they are.  Anecdotally, I think I've probably seen 'cls' used more often in the context of metaclasses - the instance of a metaclass is a class so this would seem to make sense.  PEP 8 also does not provide guidance specific to metaclasses.  Quite apart from that, I wonder whether a type checker should produce errors for stylistic peculiarities.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That it should accept `cls` as the name of the first argument.

**Screenshots or Code**
```py
class Foo(type):
    def bar(cls):
        ...
```

**VS Code extension or command-line**
extension, 1.1.12

**Comment by erictraut** [comment]
PEP 8 is pretty unequivocal:
```
Always use self for the first argument to instance methods.
Always use cls for the first argument to class methods.
```
It doesn't carve out exceptions for meta classes.

Do you see any downside to conforming with this convention? Most languages that support OOP support keywords for the instance (e.g. "this" in JavaScript, C# or C++). It's unfortunate that Python didn't carve out a keyword and make the "self" and "cls" parameter implicit. In the absence of built-in support, having a strong naming convention seems like a good thing. 

I think you could maybe make an argument for this being a warning rather than an error — or making it configurable. However, I don't think it makes sense to make an exception for metaclasses.

**Comment by erictraut** [comment]
I think I'll close this for now. If I receive feedback from additional users, I'll reconsider.

**Comment by layday** [comment]
I disagree that PEP 8 is unequivocal on this matter - I think it is perhaps more likely that the authors did not think of metaclasses when they wrote that section.  Python linters like Pyflakes, Flake8 and PyLint converged on 'cls' around the time GVR expressed his preference on the [python-dev](https://mail.python.org/pipermail/python-dev/2018-January/151986.html) mailing list.  I don't use any of those a linter, but people who do will have to silence either Pyright or the linter.

**Comment by erictraut** [comment]
The stdlib type stubs from typeshed are inconsistent on this matter. For example, ABCMeta uses `cls` but EnumMeta users `self`.

I think you've made a good point about Pyright conflicting with linters. I can think of several fixes:
1. Disable pyright's check for `self` and `cls` for methods within any class that appears to be a metaclass (i.e. derives from `type`).
2. Make pyright's checks for `self` and `cls` configurable, retaining the current behavior as default but allowing users to disable or change from an error to a warning if they want.
3. Completely remove the check from Pyright (following the logic that this is the domain of a linter, not a type checker).

@jakebailey @MikhailArkhipov @brettcannon Do you have any thoughts on this one?

**Comment by layday** [comment]
A fourth option would be to allow either 'cls' or 'self' in metaclasses but not anything exotic like 'this' or 'whoknowswhat'.

(Interestingly, the `EnumMeta` implementation uses both `cls` and `self` arbitrarily.)

**Comment by brettcannon** [comment]
So PEP 8 predates metaclasses by a lot, so don't assume it was ever updated to explicitly for metaclasses. 😉 

As for this specific case, it's a bit tricky since metaclasses can be viewed as classes in one regard and as instances in another since they are "classes for classes".

Personally, I would either:
1. Leave it to linters since this technically a stylistic thing (hence being in PEP 8); typos should get picked up via detection of names that are seemingly unbounded at reference time
2. If you want to keep it then take the simple route and accept `self` or `cls` anywhere

But personally my suggestion would be 1 from the two options and leave it to linters.

**Comment by erictraut** [comment]
Thanks for the advice Brett. I'm somewhat hesitant to remove this because it has saved me so much time and grief when I forget to add the "self" parameter. I think I'll make it configurable, default it to "warning", and relax it for metaclasses. This is a combination of several of the options above.

**Comment by erictraut** [comment]
This is now implemented in 1.1.14, which I just published. I added a new configuration setting "reportSelfClsParameterName".
